src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs

```diff
@@ -854,5 +854,49 @@ public static bool IsEnumType(this ITypeSymbol type)
         {
             return type.IsValueType && type.TypeKind == TypeKind.Enum;
         }
+
+        public static bool? IsMutableValueType(this ITypeSymbol type)
+        {
+            if (type.IsErrorType())
+            {
+                return null;
+            }
+
+            if (type.TypeKind != TypeKind.Struct)
+            {
+                return false;
+            }
+
+            if (type.IsNullable())
+            {
+                // Nullable<T> can only be mutable if T is mutable. This case ensures types like 'int?' are treated as
+                // immutable.
+                var typeArguments = type.GetTypeArguments();
+                if (typeArguments.Length != 1)
+                {
+                    return null;
+                }
+
+                if (typeArguments[0].IsNullable())
+                {
+                    // Recursion prevention. This is not a valid type anyway since the T in Nullable<T> cannot itself be
+                    // nullable.
+                    return null;
+                }
```

Comment:
the only recursion that would really be concerning would be some sort of infinite nullable type.  But... if we had that... lots of our system would likely break anyways :)  

you could also make this more simple by just doing:

```
type = typeArgument[0];
```

Then you don't need recursion at all.  You just fall through to the bottom code here, using hte inner type parameter without any recursion.

Reply:
Thanks, applied this simplification in 89849ea3bbe9af28b16191d6898e8c60b991b124